Contentful l10n support - all-in-one PR

stevejalim opened the pull request:
## Description

This changeset contains all of the l10n work for out Contentful integration and the VRC that ~is~ was featured in PR #11022 (which itself is built on PR #11002), plus some extra minor test coverage.

(This was all put into one cut following a direct request/suggestion - it was previously three separate, smaller, PRs, chained together.)

As such, merging all the PR descriptions into one, this changeset:
* Extends the current pattern of storing Contentful "pages" (Entry records and their children) as Django ContentfulEntry records, setting the appropriate locale field.
*  Displays the locale-specific content the VRC in a way that fits with exisitng Bedrock behaviour:
    * Content for which we have a specific loclized version is available if requested.
      * The locale switcher in the footer is populated with the appropriate set of locales that content is available in (be it the VRC listing page or an article page.
    * We will consider a locale to be active when there is at least certain, configurable percentage of content available in that locale compared to the default en-US locale. If the percentage is below the threshold, we act like the content is not yet in the system.  
      * The default threshold percentage for production is currently set at 60% - this was just a gut decision to get us moving
      * When `DEV=True` the threshold percentage is set very low - 1% - so that content in any published locale is shown pretty much all the time. We can experiment to see if this is has any rough corners IRL.
## Issue / Bugzilla link

Resolves #11002 (backend) 
Resolves #11022 (frontend)
Resolves #10867 (the Epic)


## Testing

### BACKEND

## Testing - assuming local Docker usage

* Set `CONTENTFUL_ENVIRONMENT=sandbox` in your  `.env`. 
* Restart your `app` container
* `make shell` 
* `./manage.py update_contentful --force`

The sandbox env is set up to have just three locales enabled (`en-US`, `de-DE`, `fr-FR`) and two of the VRC pages also have mock German translations (en-US text with "GERMAN" written above, and one of these pages also has a French translation. All should sync down to your local database, which you can check in a Django shell:

```
$ ./manage.py update_contentful --force
Running forced update from Contentful data
Pulling from Contentful
Page connectHomepage:58YIvwDmzSDjREDACTED deemed not syncable for de-DE    # Homepages only synced with no-op en-US locale
Page connectHomepage:4k3CxqZGjxXOREDACTED deemed not syncable for de-DE
Page connectHomepage:58YIvwDmzSDjREDACTED deemed not syncable for fr-FR
Page connectHomepage:4k3CxqZGjxXOREDACTED deemed not syncable for fr-FR
Updating existing ContentfulEntry for connectHomepage:en-US:58YIvwDmzSDjREDACTED  # ie, synced here.
Updating existing ContentfulEntry for connectHomepage:en-US:4k3CxqZGjxXOREDACTED
Updating existing ContentfulEntry for pagePageResourceCenter:en-US:1dQAhsz5XnrX6IREDACTED
[...]
Problem with page:3bKxRzCv6kHjCAREDACTED -> <class 'Exception'>: No 'Entry' detected for page   # bad data in sandbox
Updating existing ContentfulEntry for pagePageResourceCenter:en-US:IORNDWEoanCkQREDACTED
Updating existing ContentfulEntry for pagePageResourceCenter:en-US:5wIcW1esdWIK6REDACTED
Locale mismatch on pageHome:1XWEqmSvQQF6eREDACTED -> en-US vs Untitled page - 📄 Page: Home   # bad data in sandbox
Creating new ContentfulEntry for pagePageResourceCenter:en-US:2LEDujIfxanZKMREDACTED
Locale mismatch on pageHome:60YAs15SYSKKnjREDACTED -> en-US vs Untitled page - 📄 Page: Home   # bad data in sandbox
Locale mismatch on pageHome:1HercJWjhIEm4SREDACTED -> en-US vs Untitled page - 📄 Page: Home   # bad data in sandbox
Creating new ContentfulEntry for pagePageResourceCenter:de-DE:1dQAhsz5XnrX6IREDACTED
Creating new ContentfulEntry for pagePageResourceCenter:de-DE:4Gobq3NdbOCaaaREDACTED
No content for 1JxYJUTGphwTNRREDACTED for de-DE
No content for 3gikThlF98f2heREDACTED for de-DE
[...]
Creating new ContentfulEntry for pagePageResourceCenter:fr-FR:1dQAhsz5XnrX6IREDACTED
No content for 4Gobq3NdbOCaaaREDACTED for fr-FR
[...]
No content for 1HercJWjhIEm4SREDACTED for fr-FR
Entries to be deleted: <QuerySet []>
Deleted by _detect_and_delete_absent_entries: (0, {'contentful.ContentfulEntry': 0})
Done. Added: 4. Updated: 8. Deleted: 0. Errors: 4
```

##### Things to experiment with

Fire up Django shell and try out these fun things! They're most easily checked by running ORM queries for `bedrock.contentful.models.ContentfulEntry`

* Archive an entire multi-locale page and check what happens when you force-sync locally. Do the relevant pages get deleted? Do the remaining pages persist?
* Unarchive the archived page and re-sync - all the pages will be back
* Edit a multi-locale Compose page in Contentful and remove the SEO and main content entries/children for a particular locale. Force resync - the speciifc zapped locale should be deleted but the pages other locales should remain
* Restore/re-link the missing data from the previous step (bear in mind how we model SEO objects may change once we get data back from translation, so go with the triple deck of SEO objects for now). Re-sync. Nice to have the data back, yep?

#### FRONTEND / VRC changes

* Force-sync the sandbox content, as described for the Backend manual tests 
* Locally, visit the VRC at http://localhost:8080/en-US/products/vpn/resource-center/
* Check the locale switcher in the footer - confirm you can switch to German (two fake-example articles) and French (one fake-example article)
* Explore those links on the DE and FR pages - confirm you don't get booted back to the default en-US
* Try changing the slug so that the listing page that is available in one locale is not available for the other locale  - eg set the locale to `ru` or `ja` or `es-MX` - ~you should get 404ed~ you should get sent to the en-US version of the page as the fallback
* Try changing the slug for a known locale (eg `de`) to one that goes not exist (eg `xx`). You will get a 404 for listing and article pages
* Try changing the locale for a known page to a different locale which lacks a translation for that page - eg change http://localhost:8080/de/products/vpn/resource-center/the-difference-between-a-vpn-and-a-web-proxy/ to a `fr` locale. You should get sent to the default `en-US` locale page, because we do have content for that slug in that fallback locale
* Confirm that even if we have a viable locale, asking for an article with an uknown slug will, of course, 404: http://localhost:8080/de/products/vpn/resource-center/asdasdaa/

pmac commented on bedrock/settings/base.py:
```diff
@@ -1039,6 +1040,22 @@ def get_app_name(hostname):
 CONTENTFUL_NOTIFICATION_QUEUE_SECRET_ACCESS_KEY = config("CONTENTFUL_NOTIFICATION_QUEUE_SECRET_ACCESS_KEY", default="", raise_error=False)
 CONTENTFUL_NOTIFICATION_QUEUE_WAIT_TIME = config("CONTENTFUL_NOTIFICATION_QUEUE_WAIT_TIME", default="10", parser=int, raise_error=False)
 
+CONTENTFUL_HOMEPAGE_LOOKUP = {
```
Perhaps we should use these in the home page view as well, just to make sure this remains accurate.

stevejalim replied:
We do now use them in `bedrock.mozorg.views.home_view`, if that's what you mean:

eg https://github.com/mozilla/bedrock/pull/11061/files#diff-6cfdf8d97afa7adfe1b9d1ff09b2c25192abf7a08f60e6bd8d3efe6497625ad5R135

```
ctx.update(ContentfulEntry.objects.get_page_by_id(content_id=settings.CONTENTFUL_HOMEPAGE_LOOKUP["en-US"]))
```

pmac replied:
Doh! Meant to delete this comment before posting the review. I did find that. Thanks.

pmac commented on bedrock/products/views.py:
```diff
@@ -140,17 +145,23 @@ def _filter_articles(articles_list, category):
+    # threshold percentage, we should send you to the default locale
+    if locale not in active_locales:
+        _path = reverse("products.vpn.resource-center.landing")
+        return redirect(f"/{DEFAULT_LOCALE}{_path}")
```
you could redirect to the URL without a locale and let bedrock decide which is best for the user.

stevejalim replied:
I just started amending this but then reverted. I _think_ this went in as an explicit fallback to `en-US`/`DEFAULT_LOCALE` because if you reach here you are trying to use a locale which is not active, and if there's a possibility that if your browser is _insisting_ on a that locale, you'll just end up in a redirect loop seeking an unavailable locale.

I could be wrong - there may be a gap in my understanding of Bedrock's automatic locale selection that avoids the redirect loop problem, but if not, I think this is going to give us a usable fallback that's avoids the risk of that loop.

What d'ya think?

pmac replied:
Maybe we need to think this through a bit more. Bedrock will choose the best active locale, but that happens in the `l10n_utils.render` function, which is called after this view. So maybe the best option if we find this case is to just call render with the active locales and let it handle the redirect?

stevejalim replied:
Yeah that sounds a lot simpler. Will sort it 🎸

pmac commented on bedrock/contentful/utils.py:
```diff
@@ -0,0 +1,68 @@
+        content_type=content_type,
+        classification=classification,
+    )
+    if slug:
```
Just making sure I'm following correctly. This is used for the main view of the resource center when called without a slug. Then when it is called with a slug, the percentages below should all be 100 since each active locale should have 1 article? Is that right?

stevejalim replied:
Looking back at this now, I think this can be made easier with an early return if we have a slug - once we call
```
all_locales = ContentfulEntry.objects.filter(**kwargs).values_list("locale", flat=True)
```
then that's all the locales for that slug + content type + classification so they are all valid. There's no need to do the extra legwork of the threshold check at all.

Will update things now

stevejalim replied:
(This is what comes from overloading something build for A to do A+B without stepping back to see the bigger picture 🤦 )

pmac reviewed (commented):
Looking good so far. I just have some questions, and then I think we can get this merged.

stevejalim commented on bedrock/contentful/api.py:
```diff
@@ -23,10 +23,24 @@
-# Bedrock to Contentful locale map
-LOCALE_MAP = {
+# Bedrock and Contentful have slightly differing locale lists.
+BEDROCK_TO_CONTENTFUL_LOCALE_MAP = {
```
Ah, I'll expand the code comment here. These are the mappings where the Contentful code and the Bedrock code don't directly match, so we need to fix them, rather than this being the _only_ ones we support. eg Bedrock has `pl` but Contentful models that as `pl-PL` so we 'flatten' it to match what Bedrock currently uses internally. 

Looking back at the list, the majority are in the pattern of the suffix just being the uppercase version of the prefix. 

`ms-MY` has been mapped to `ms` but looking at it cold, we should double-check that internally we are using that as Malaysian Malay and not anything to do with Monserrat.

stevejalim replied:
Looking at `ms-MY` <-> `ms` - this does make sense as the correct locale transformation (`ms` -> Malaysian Malay in production whereas`my` -> Burmese, for Myanmar)

stevejalim commented on bedrock/products/tests/test_views.py:
```diff
@@ -286,18 +286,43 @@ def test_simple_get__for_valid_locale_with_enough_content(self, render_mock):
             ],
         )
 
+    def test_simple_get__for_invalid_locale(self, render_mock):
```
These tests were swapped in order, to make context easier when reading them all cold in the future

stevejalim commented:
@pmac hi - I've addressed all the current questions and tweaked code where needed. I'd def welcome your thoughts on the changes, particularlyl the early return pattern I've added. 

The overall idea sits fine with me, but the implementation here still feels a bit messy/not-quite-right (I've leaned towards pre-defining the params for render early, to reduce risk, rather than write them out twice, but it doesn't feel very idiomatic). 

Alternative angles: 
* If we refactored to CBVs for this, we might clean up some of the icky bits, but then the cost would be the opaqueness of CBVs. 
* Pre-canning the return function in a closure and relying on pass-by-reference to keep the `ctx` variable up to date has the same drawbacks of just defining them early.

Or maybe it's fine/intelligible enough as-is... I'll re-look at it in the morning with a fresh brain

stevejalim commented:
@pmac Ok, just added the cleaner approach with the refactored-to-be-separate redirect function. Take a look whenever you get time, but this could be enough for now.

Pushing to the integraiton-tests branch now, just in case there's a test that doesn't like the minor refactoring of `l10n_utils.render`, too

pmac commented on bedrock/products/views.py:
```diff
@@ -141,48 +143,48 @@ def _filter_articles(articles_list, category):
+    requested_locale = l10n_utils.get_locale(request)
+
+    if requested_locale not in active_locales:
+        return l10n_utils.redirect_to_best_locale(
```
Love it. Nice and clean and clear.

pmac commented on bedrock/settings/base.py:
```diff
@@ -1046,6 +1047,22 @@ def get_app_name(hostname):
+
+CONTENTFUL_LOCALE_ACTIVATION_PERCENTAGE = config(
+    "CONTENTFUL_LOCALE_ACTIVATION_PERCENTAGE",
+    default="60",
```
We should do the `if DEV` bit here so that demos or local dev can set this even in DEV mode. e.g.

```python
CONTENTFUL_LOCALE_ACTIVATION_PERCENTAGE = config(
    "CONTENTFUL_LOCALE_ACTIVATION_PERCENTAGE",
    default="1" if DEV else "60",
    parser=float,
)
```

pmac reviewed (approved):
I made one more suggestion, but after that I think we're good to merge. So merge when ready. 👏
